File: sdks/node/tests/ssh-local.integration.test.ts
import { mkdtempSync, readFileSync, rmSync } from "node:fs";
import { execFileSync } from "node:child_process";
import { expect, test } from "vitest";
import { getJsBoxlite } from "../lib/native.js";

test("SSH local native configure status disable and restart", async () => {
  const keys = mkdtempSync("/tmp/boxlite-node-ssh-");
  const runtime = getJsBoxlite().withDefaultConfig();
  const box = await runtime.create({ image: "alpine:3.19", autoRemove: false });
  try {
    for (const name of ["host", "user"])
      execFileSync("ssh-keygen", [
        "-q",
        "-t",
        "ed25519",
        "-N",
        "",
        "-f",
        `${keys}/${name}`,
      ]);
    const config = {
      listenAddress: "0.0.0.0:2222",
      hostPrivateKey: readFileSync(`${keys}/host`, "utf8"),
      accounts: [
        {
          login: "alice",
          authorizedKeys: [readFileSync(`${keys}/user.pub`, "utf8")],
        },
      ],
    };
    expect((await box.ssh.status()).generation).toBe(0n);
    expect((await box.ssh.configure(config)).enabled).toBe(true);
    expect((await box.ssh.configure(config)).generation).toBe(2n);
    expect((await box.ssh.disable()).enabled).toBe(false);
    await box.stop();
    const fresh = (await runtime.get(box.id))!;
    expect((await fresh.ssh.status()).generation).toBe(0n);
  } finally {
    await runtime.remove(box.id, true);
    rmSync(keys, { recursive: true, force: true });
  }
});

File: sdks/go/ssh_local_integration_test.go
//go:build boxlite_dev

package boxlite

import (
	"context"
	"os"
	"os/exec"
	"path/filepath"
	"testing"
	"time"
)

func TestSSHLocalLifecycle(t *testing.T) {
	rt := newTestRuntime(t)
	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
	defer cancel()
	keys := t.TempDir()
	for _, name := range []string{"host", "user"} {
		if output, err := exec.CommandContext(ctx, "ssh-keygen", "-q", "-t", "ed25519", "-N", "", "-f", filepath.Join(keys, name)).CombinedOutput(); err != nil {
			t.Fatalf("keygen: %v %s", err, output)
		}
	}
	host, err := os.ReadFile(filepath.Join(keys, "host"))
	if err != nil {
		t.Fatal(err)
	}
	key, err := os.ReadFile(filepath.Join(keys, "user.pub"))
	if err != nil {
		t.Fatal(err)
	}
	box, err := rt.Create(ctx, "alpine:3.19", WithAutoRemove(false))
	if err != nil {
		t.Fatal(err)
	}
	defer func() {
		cleanupCtx, cleanupCancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cleanupCancel()
		if err := rt.ForceRemove(cleanupCtx, box.ID()); err != nil {
			t.Errorf("remove SSH test box: %v", err)
		}
	}()
	defer box.Close()
	ssh, err := box.SSH()
	if err != nil {
		t.Fatal(err)
	}
	defer ssh.Close()
	config := SSHConfig{ListenAddress: "0.0.0.0:2222", HostPrivateKey: string(host), Accounts: []SSHAccount{{Login: "alice", AuthorizedKeys: []string{string(key)}}}}
	status, err := ssh.Status(ctx)
	if err != nil || status.Generation != 0 {
		t.Fatalf("initial: %v %v", status, err)
	}
	status, err = ssh.Configure(ctx, config)
	if err != nil || !status.Enabled {
		t.Fatalf("configure: %v %v", status, err)
	}
	status, err = ssh.Configure(ctx, config)
	if err != nil || status.Generation != 2 {
		t.Fatalf("reconfigure: %v %v", status, err)
	}
	status, err = ssh.Disable(ctx)
	if err != nil || status.Enabled {
		t.Fatalf("disable: %v %v", status, err)
	}
}
